### PR TITLE
Rewrite `next` branch to fix all our sharp edges

### LINKED ISSUE #55: Warning while compiling
```
warning: src/scanner.cc:179:17: warning: result of comparison of constant -1 with expression of type '(anonymous namespace)::TokenType' is always true [-Wtautological-constant-out-of-range-compare]
warning:       if (token != TOP_LEVEL && token != OPEN_BRACE) {
warning:           ~~~~~ ^  ~~~~~~~~~
warning: 1 warning generated.
```

Using a `Cargo.toml` with a tree-sitter-r dependency pointed at a `{ path = 'local/path' }`

### LINKED ISSUE #62: "..i" node name gives problems with tree-sitter-node
The `..i` token defined here: https://github.com/r-lib/tree-sitter-r/blob/1e9eaa9af98c9ac92e3d1e1713a453eef488fe36/grammar.js#L304

Causes an error when this parser is loaded in `tree-sitter-node`, because it creates a class for each syntax node and a dot (`.`) is illegal in a class name.

Is it possible to use a different name (without special characters) for this token?

### LINKED ISSUE #65: Strange tree-sitter parse tree in the `next` branch
This code is interpreted as one large call with arguments `library()`, `1`, `2`, `3`.

```r
library(dplyr::)

library()

1
2
3
```

This is a bit surprising could we do better here?

### LINKED ISSUE #77: `externals` should actually have unique names, not alias to string literals
In https://github.com/r-lib/tree-sitter-r/pull/47, @lionel- simplified our externals by mapping them to string literals like `"]"` rather than unique names like `$._closing_bracket`.

I've determined that we really do want them to be unique names, as otherwise any usage of `"]"` in the grammar can accidentally match in the _internal_ scanner if we've refused to handle a `]` token in the _external_ scanner.

i.e. the TLDR is that we _only_ want the `]` to match if our external scanner says we found one, but right now that isn't the case.

Consider `x[y[[2]`. That closing `]` is invalid and should be considered an unmatched `]` to be matched by our `]:` rule for unmatched closing brackets.

However, it currently gets matched to the first opening `[`, essentially 'closing' it.

```
(program
  ([
    (identifier)
    (arguments
      (argument
        ([[
          (identifier)
          (arguments
            (argument
              (float))))))))
```

This is due to our external being called `"]"` and then our usage of it in:

```
_subset_arguments:  $ => prec.right(seq("[",  repeat(choice($.comma, $._semicolon, field("argument", $.argument))), optional("]"))),
```

We refuse to deal with the `]` in the above example because our external scanner correctly detects that it isn't currently valid, but the internal scanner then gets a chance to run and ends up matching it to `optional("]")` due to that being a simple string literal.

If we replace `optional("]")` with `optional($._closing_bracket)` then we'd get what I'd expect:

```
(program
  ([
    (identifier)
    (arguments
      (argument
        ([[
          (identifier)
          (arguments
            (argument
              (float))
            (argument
              (]))))))))
```

### LINKED ISSUE #81: Incorrect `|>` , `+` precedence combination
```r
x |> foo() + baz()
```

gives

```
(program 
  (|> 
    lhs: (identifier) 
    rhs: (+ 
      lhs: (call 
        (identifier) 
         arguments: (arguments)) 
      rhs: (call 
         (identifier) 
         arguments: (arguments)))))
```

should be

```
(program
  (+
    (|>
      (identifier)
      (call
        (identifier)
        (arguments)))
    (call
      (identifier)
      (arguments)))
```

i.e. pipe happens first internally, then `+` happens at the end

### LINKED ISSUE #82: `$` and `@` should not allow arbitrary expressions on the RHS
Only strings and symbols
https://github.com/wch/r-source/blob/36008873fb8ca2af3bdaaff418dbade5f7bce118/src/main/gram.y#L512-L515

```r
foo${hi}
```

gives

```
(program ($ lhs: (identifier) rhs: ({ body: (identifier))))
```

but really that should probably error

### PR DESCRIPTION
**Pull request by DavisVaughan**
Closes #82 
Closes #81 
Closes #77 
Closes #65 
Closes #55 
Closes #62 
Closes #64 

I actually think the diff of `grammar.js` is not horrible if you want to look at that

I have added _a lot_ of tree sitter tests too, which will also get brought into the R package in the future

There is a big focus on adding `NOTE`s any time we have to do anything weird with precedence, as that is typically the weirdest part about the grammar and it is hard to know when a change was made because it is in line with R vs something that is required to make it work for tree sitter

### CODE REVIEW

File: grammar.js [hunk 1]
```diff
@@ -209,149 +189,262 @@ module.exports = grammar({
...
+      optional($._close_parenthesis)
     )),
 
     // Function calls and subsetting.
-    call: $ => prec.right(16, seq($._expression, field("arguments", alias($._call_arguments,    $.arguments)))),
-    "[":  $ => prec.right(16, seq($._expression, field("arguments", alias($._subset_arguments,  $.arguments)))),
-    "[[": $ => prec.right(16, seq($._expression, field("arguments", alias($._subset2_arguments, $.arguments)))),
-
-    // Dummy rule; used for aliasing so we can easily search the AST.
-    arguments: $ => $._expression,
```
Comment (DavisVaughan) [inline grammar.js:283]: Neat side note, we don't actually need this, we can do `alias($.subset_arguments, $.arguments)` in the `subset:` rule instead

File: grammar.js [hunk 2]
```diff
@@ -209,149 +189,262 @@ module.exports = grammar({
...
+    repeat_statement: $ => withPrec(PREC.FUNCTION_OR_LOOP, seq(
       "repeat",
       repeat($._newline),
       optional(field("body", $._expression))
     )),
 
     // Blocks.
-    "{": $ => prec.right(seq(
-      "{",
+    braces: $ => withPrec(PREC.BLOCK, seq(
```
Comment (kevinushey) [inline grammar.js:261]: Maybe `braced_expression`? `block`? Just to emphasize that this captures an expression within `{}`.

File: grammar.js [hunk 3]
```diff
@@ -209,149 +189,262 @@ module.exports = grammar({
...
+      $._open_brace,
       repeat(field("body", choice($._expression, $._semicolon, $._newline))),
-      optional("}")
+      optional($._close_brace)
     )),
 
-    "(": $ => prec.right(seq(
-      "(",
+    parentheses: $ => withPrec(PREC.BLOCK, seq(
```
Comment (kevinushey) [inline grammar.js:267]: Maybe `paren_expression`? Similar to above, emphasize that this captures an expression within `()`.

**Review by kevinushey — APPROVED** [review]
LGTM; very nice work!

File: grammar.js (hunk 2 above)
Comment (DavisVaughan) [reply grammar.js]: Went with `braced_expression` and `parenthesized_expression`, as those use the same tense and match `parenthesized_expression` in the python grammar

**Pull request merged**